packages/steveo/src/runtime/workflow-state.ts

```diff
@@ -0,0 +1,54 @@
+
+// TODO: Write Prisma persistence
+
+/**
+ * Tracks the state of a workflow execution.
+ */
+export interface WorkflowState {
+
+  /**
+   * The unique key of the workflow execution instance.
+   */
+  flowId: string;
+
+  /**
+   * The time the flow execution began.
+   * Diagnostics and auditing.
+   */
+  started: Date;
+
+  /**
+   * The current step being executed in this flow.
+   * Will increment under "happy path" flow where we are advancing through
+   * the steps, or decrementing if the flow has failed and rolling back.
+   */
+  current: number;
+
+  /**
+   * The initial value that started the flow execution.
+   * The engine will pass this value to the first step.
+   */
+  initial: unknown;
+
+  /**
+   * Step state tracking.
+   * The next step in the chain will be executed with the result
+   * of the previous step as its argument.
+   */
+  results: unknown[];
+
+  /**
+   * Diagnostic and auditing tracking of the step a flow execution failed
+   */
+  failedStep?: number;
+
+  /**
+   * Diagnostic and auditing tracking of the failed error message
+   */
+  failedErrMsg?: string;
+
+  /**
+   * Diagnostic and auditing tracking of the failed error callstack
+   */
+  failedErrStack?: string;
```

Comment:
This assumes that only one step can fail. The rollback step could also fail. What's stored here might be useful for flow control, but for debugging I think we need to have an object in the `results` that covers both success and failure. That way you can find out which steps failed and what the errors were

Reply:
Agreed, I'll turn this into more of a JSON-esq capture of errors

Reply:
Updated the contract, will do the storage implementation this afternoon

https://github.com/ordermentum/steveo/pull/800/commits/1e1234041b1cf44adb3516b7f5c760972f5da4f7